### PR TITLE
KAFKA-16339: [4/4 KStream#flatTransformValues] Remove Deprecated "transformer" methods and classes

### PR DESCRIPTION
**Pull request by fonsdant**

### CODE REVIEW

**Comment by fonsdant** [comment]
Hi, @mjsax! I have a question for this PR. I'm wondering if `shouldNotAllowNullStoreNameOnProcessValues` were not duplicated as `shouldNotAllowNullStoreNamesOnProcessValues` (note the little "s" in "Names"). What do you think?

**Comment by mjsax** [comment]
Looking into the test code, it seem the `s` just goes somewhere else? There is `shouldNotAllowNullStoreNameOnProcess` and `shouldNotAllowNullStoreName[s]OnProcess`, one taking `(String) null` and the other taking a `(String[]) null`, so it seems we are fine?

**Comment by fonsdant** [comment]
> Looking into the test code, it seem the `s` just goes somewhere else? There is `shouldNotAllowNullStoreNameOnProcess` and `shouldNotAllowNullStoreName[s]OnProcess`, one taking `(String) null` and the other taking a `(String[]) null`, so it seems we are fine?

Thanks, @mjsax! Alright, just a confirmation.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 1]
```diff
@@ -1204,75 +1201,6 @@ private <VO, VR> KStream<K, VR> doStreamTableJoin(final KTable<K, VO> table,
...
-    }
-
-    @Override
-    @Deprecated
-    public <VR> KStream<K, VR> flatTransformValues(final org.apache.kafka.streams.kstream.ValueTransformerSupplier<? super V, Iterable<VR>> valueTransformerSupplier,
-                                                   final Named named,
-                                                   final String... stateStoreNames) {
-        Objects.requireNonNull(valueTransformerSupplier, "valueTransformerSupplier can't be null");
-        return doFlatTransformValues(
-            toValueTransformerWithKeySupplier(valueTransformerSupplier),
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:1223]: I think we can also remove `toValueTransformerWithKeySupplier(...)` ? Should be unused now?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 2]
```diff
@@ -1204,75 +1201,6 @@ private <VO, VR> KStream<K, VR> doStreamTableJoin(final KTable<K, VO> table,
...
-    }
-
-    private <VR> KStream<K, VR> doFlatTransformValues(final ValueTransformerWithKeySupplier<? super K, ? super V, Iterable<VR>> valueTransformerWithKeySupplier,
-                                                      final Named named,
-                                                      final String... stateStoreNames) {
-        Objects.requireNonNull(stateStoreNames, "stateStoreNames can't be a null array");
-        for (final String stateStoreName : stateStoreNames) {
-            Objects.requireNonNull(stateStoreName, "stateStoreNames can't contain `null` as store name");
-        }
-        ApiUtils.checkSupplier(valueTransformerWithKeySupplier);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:1252]: I think we can also remove this `checkSupplier(...)` overload from `ApiUtils` ?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Made a pass. Think we can remove a little more code, which is unused now.

**Comment by mjsax** [comment]
@fonsdant -- This PR need a rebase to resolve conflicts. Would also be great if you could also address the two minor comments from the review, so we can merge this PR.

**Comment by fonsdant** [comment]
Thanks, @mjsax! Sorry for the delay this week. I can push the changes by tomorrow.

**Comment by mjsax** [comment]
Thanks for the PR. Merged to `trunk`.

**Comment by mjsax** [comment]
@fonsdant -- Would you be interested to help us updating the docs?

Eg https://kafka.apache.org/39/documentation/streams/developer-guide/dsl-api.html#applying-processors-and-transformers-processor-api-integration -- there might be other parts, too, which we need to update.

It would also be good to maybe add a guide (ie, some simple examples) how to rewrite old code for the new `process()` method we have?

Besides `[flat]Transform[Values]` stuff, we did remove a lot of other API for 4.0. If you are interested to help updating the docs for these, too, would be great.

Don't feel obligated. Just let us know. -- If you don't have interest, we can also work through it on our side.

**Comment by fonsdant** [comment]
@mjsax, of course! Thanks for mentioning me! Have we already a task for this or should I create a minor PR?

**Comment by mjsax** [comment]
MINOR PR is fine -- if you do a dedicated PR only for `[flat]Transform[Values]`, you could also use KAFKA-16339.

**Pull request merged**